baybe/campaign.py

```diff
@@ -222,6 +268,100 @@ def add_measurements(
             numerical_measurements_must_be_within_tolerance,
         )
 
+    def _mark_as_measured(
+        self,
+        measurements: pd.DataFrame,
+        numerical_measurements_must_be_within_tolerance: bool,
+    ) -> None:
+        """Mark the given elements of the discrete subspace as measured.
+
+        Args:
+            measurements: A dataframe containing parameter configurations to be
+                marked as measured.
+            numerical_measurements_must_be_within_tolerance: See
+                :func:`baybe.utils.dataframe.fuzzy_row_match`.
+        """
+        idxs_matched = fuzzy_row_match(
+            self.searchspace.discrete.exp_rep,
+            measurements,
+            self.parameters,
+            numerical_measurements_must_be_within_tolerance,
+        )
+        self._searchspace_metadata.loc[idxs_matched, _MEASURED] = True
+
+    @singledispatchmethod
+    def toggle_discrete_candidates(  # noqa: DOC501
+        self,
+        constraints: Collection[DiscreteConstraint] | pd.DataFrame,
+        exclude: bool,
+        complement: bool = False,
+        dry_run: bool = False,
+    ) -> pd.DataFrame:
+        """In-/exclude certain discrete points in/from the candidate set.
+
+        Args:
+            constraints: A filtering mechanism determining the candidates subset to be
+                in-/excluded. Can be either a collection of
+                :class:`~baybe.constraints.base.DiscreteConstraint`s or a dataframe.
+                For the latter, see :func:`~baybe.utils.dataframe.filter_df`
+                for details.
+            exclude: If ``True``, the specified candidates are excluded.
+                If ``False``, the candidates are considered for recommendation.
+            complement: If ``False``, the filtering mechanism is used as is.
```

Comment:
Isn't this first sentence redundant? I think it can be removed and this docstring can be reformulated in the style `Setting this flag inverts the filtering mechanism such that....`. Alternatively, one could say `If False, then the filtering is performed as described by constraints and exclude. If True, the ....`, but I'd prefer any of these alternatives over the current one.

Reply:
But also on the plus side, I think the naming of the attributes and the way this function now works is in general significantly clearer 👍

Reply:
I think the second option you mentioned is the same I had just in different words, tbh. And the first one is fine for me, I just started it with `If True, ...` to keep the style consistent. But that effectively resulted pretty much in simply dropping my original first sentence – which is fine for me 👍🏼